src/ComponentProvider.ts

```diff
@@ -0,0 +1,95 @@
+import type { Schema, Message, AsyncAPIDocument, ChannelParameter } from '@asyncapi/parser';
+
+/**
+ * This Singleton class will provide all sorts of data for optimizers.
+ *
+ * @public
+ */
+export class ComponentProvider {
+  messages = new Map<string, Message>();
+  schemas = new Map<string, Schema>();
+  parameters = new Map<string, ChannelParameter>();
+
+  constructor(private document: AsyncAPIDocument) {
+    this.scanChannels();
+    this.scanComponents();
+  }
+
+  private scanChannels = (): void => {
+    const channels = this.document.channels();
+    for (const channelName in channels) {
+      const path = `channels.${channelName}`;
+      const channel = this.document.channel(channelName);
+
+      if (channel.hasPublish()) {
+        const message = channel.publish().message();
+        const currentPath = `${path}.publish.message`;
+        this.scanMessage(currentPath, message);
+        this.messages.set(currentPath, message);
+      }
+
+      if (channel.hasSubscribe()) {
+        const message = channel.subscribe().message();
+        const currentPath = `${path}.subscribe.message`;
+        this.scanMessage(currentPath, message);
+        this.messages.set(currentPath, message);
+      }
+
+      if (channel.hasParameters()) {
+        const currentPath = `${path}.parameters`;
+        for (const [name, channelParameter] of Object.entries(channel.parameters())) {
+          this.parameters.set(`${currentPath}.${name}`, channelParameter);
+        }
+      }
+    }
+  }
+  private scanHeader = (path: string, headers: any): void => {
+    this.schemas.set(path, headers);
+    const headersProperties = headers.properties;
+
+    for (const [propertyName, propertySchema] of Object.entries(headersProperties)) {
+      this.schemas.set(`${path}.properties.${propertyName}`, propertySchema as Schema);
+    }
+  }
+  private scanMessage = (path: string, message: Message): void => {
+    const payload = message.payload();
+
+    if (payload) {
+      this.schemas.set(`${path}.payload`, payload);
+      const payloadProperties = payload.properties();
+      for (const [propertyName, propertySchema] of Object.entries(payloadProperties)) {
+        this.schemas.set(`${path}.payload.properties.${propertyName}`, propertySchema);
+      }
+    }
+    
+    //scanning the traits.
+    const traits = message.extension('x-parser-original-traits');
+    for (let i = 0; i < traits.length; i++) {
+      if (Object.keys(traits[i]).includes('headers')) {this.scanHeader(`${path}.traits[${i}].headers`, message.headers());}
+    }
+  }
```

Comment:
Where you scan `message.headers`? Remember that `payload` can be also also in the `traits`.

Reply:
We have a function for that. `this.scanHeader`.
I was thinking about implementing the scanning and optimization for the whole traits. But this PR has become too large for additional features. Because of that I stuck with the current features and I will take care of traits in a seperate PR.

Reply:
But I mean that you don't check the `message.headers` only `message.payload` here.

Reply:
For clarify: you do this:

```ts
    const payload = message.payload();

    if (payload) {
```

but I don't see this:

```ts
    const headers = message.headers();

    if (headers) {
```

you can define `headers` without `traits`.

Reply:
If we can have `headers` in traits and outside traits:
1) does `message.headers()` check both places and return the existing one? 

2) What if we have headers in both places at the same time? Then what happens?

Reply:
`message.headers()` function returns merged plain headers from `message.headers` field and from traits.

You should operate on the doc before optimization, because then someone can define some headers in `message.headers` property and also add additional headers in traits. You should check duplication here and here - traits you can leave for separate PR. This same situation is with `payload`. You can define payload in `message.payload` and `message.traits.*.payload`.

Reply:
@magicmatatjahu I was thinking about how to approach `headers` and `payload` optimization. I think the best approach would be to optimize the `message.headers` and `message.payload` property for now without using the `message.headers() `or `message.payload()`. then we can optimize the `message.traits.*.headers` and `message.traits.*.payload` as a part of `traits` optimization.
what do you think?

Reply:
@magicmatatjahu  Done. I have taken a different approach for handling this issue please take a look.